### PR TITLE
feat(rpc): eth_sign*

### PR DESCRIPTION
**Pull request by fkrause98**
### Endpoints to be added:
- [x] eth_sign
- [x] eth_signTypedData

ref: https://github.com/paradigmxyz/reth/issues/1648

### CODE REVIEW

File: crates/rpc/rpc/src/eth/signer.rs
Comment (mattsse) [inline crates/rpc/rpc/src/eth/signer.rs:38]: do we need pub crate access here?
imo should be private.

File: crates/rpc/rpc/src/eth/signer.rs
Comment (mattsse) [inline crates/rpc/rpc/src/eth/signer.rs:54]: we need additional error types here:

this is what geth returns if no account found:

https://github.com/ethereum/go-ethereum/blob/ee530c0d5aa70d2c00ab5691a89ab431b73f8165/accounts/errors.go#L26-L26

File: crates/rpc/rpc/src/eth/api/sign.rs
Comment (mattsse) [inline crates/rpc/rpc/src/eth/api/sign.rs:26]: the sign RPC responses are `hex(sig)` not hex(rlp(sig))

also the response should use the `v` value and not `odd_y_parity`

looks like we need additional helper functions for Signature.

this is what anvil returns for example:
https://github.com/gakonst/ethers-rs/blob/72428503cfb0855579d47f691b48b2d65ad8195b/ethers-core/src/types/signature.rs#L65-L70
which already has the v value set.

**Review by mattsse — CHANGES_REQUESTED** [review]
ty,
some early comments

File: crates/rpc/rpc/src/eth/api/sign.rs
Comment (fkrause98) [reply crates/rpc/rpc/src/eth/api/sign.rs]: Done! [Here it is](https://github.com/paradigmxyz/reth/pull/1665/files#diff-ca37ac517ba9b4c3b5a4c91954590aebeb5fd89ae8b8d7404056b19f966b38d4R20)

File: crates/rpc/rpc/src/eth/api/sign.rs
Comment (fkrause98) [reply crates/rpc/rpc/src/eth/api/sign.rs]: Removed :)

File: crates/rpc/rpc/src/eth/api/sign.rs
Comment (fkrause98) [reply crates/rpc/rpc/src/eth/api/sign.rs]: Done!

Now only the handler returns Result<Bytes> and the sign function returns EthResult<Signature>

File: crates/rpc/rpc/src/eth/api/sign.rs
Comment (fkrause98) [reply crates/rpc/rpc/src/eth/api/sign.rs]: Ok, so, I've done a couple of changes:
- The `hex(sig)`part is being done [here](https://github.com/paradigmxyz/reth/pull/1665/files#diff-e215e50528053cd9402e7ad953d55209aa4399f5f581236bed4d582970bd793eR103). This turns the signature into a [u8;65], [like in the file you linked](https://github.com/gakonst/ethers-rs/blob/72428503cfb0855579d47f691b48b2d65ad8195b/ethers-core/src/types/signature.rs#L223)
- I've also added a From<&Signature> to Bytes as to return a Result<Bytes> on the handler.

Let me know if this is correct or if there's something else that needs fixing ^^

File: crates/rpc/rpc/src/eth/signer.rs
Comment (fkrause98) [reply crates/rpc/rpc/src/eth/signer.rs]: No, no we don't. Was testing some things and committed this by mistake, made them private again.

File: crates/rpc/rpc/src/eth/signer.rs
Comment (fkrause98) [reply crates/rpc/rpc/src/eth/signer.rs:54]: Could you go a bit more into detail on where should I add those error types? Do you mean RPCErrors or EthApi errors? Thanks :)

**Comment by fkrause98** [comment]
I've also added the eth_signedTypeData now, so opening this and waiting on some more feedback 👍 .

File: crates/rpc/rpc/src/eth/signer.rs
Comment (mattsse) [reply crates/rpc/rpc/src/eth/signer.rs]: we want these as additional variants in `EthApiError`, like `EthApiError::AccountNotFound`

File: crates/primitives/src/transaction/signature.rs
Comment (mattsse) [inline crates/primitives/src/transaction/signature.rs:103]: imo this kind of from impl should be a functions.

File: crates/primitives/src/transaction/signature.rs
Comment (mattsse) [inline crates/primitives/src/transaction/signature.rs:109]: nit: please no additional step

pass the function result directly into copy_from_slice

File: crates/primitives/src/transaction/signature.rs
Comment (mattsse) [inline crates/primitives/src/transaction/signature.rs:111]: the signature schemes uses v = 27 or 28, so what we need to do here is

```rust
 let v = u8::from(self.odd_y_parity) + 27;
 sig[64] = v;
```

File: crates/primitives/src/transaction/signature.rs
Comment (mattsse) [inline crates/primitives/src/transaction/signature.rs:122]: please remove this from impl.

File: crates/rpc/rpc/src/eth/api/sign.rs
Comment (mattsse) [inline crates/rpc/rpc/src/eth/api/sign.rs:14]:
```suggestion

impl<Client, Pool, Network> EthApi<Client, Pool, Network> {
```

File: crates/rpc/rpc/src/eth/api/sign.rs
Comment (mattsse) [inline crates/rpc/rpc/src/eth/api/sign.rs:18]: imo, it would be better to use an opaque signer error variant here (Box< dyn Error>)

File: crates/rpc/rpc/src/eth/signer.rs
Comment (mattsse) [inline crates/rpc/rpc/src/eth/signer.rs:45]:
```suggestion

    fn sign_hash(&self, hash: H256, account: Address) -> Result<Signature> {
```

File: crates/rpc/rpc/src/eth/signer.rs
Comment (mattsse) [inline crates/rpc/rpc/src/eth/signer.rs:46]: use global

File: crates/rpc/rpc/src/eth/signer.rs
Comment (mattsse) [inline crates/rpc/rpc/src/eth/signer.rs:48]: error can be replaced with variant.

function should have proper error response.

**Review by mattsse — CHANGES_REQUESTED** [review]
good progress.

please use proper errors where possible

RPC error should be reserved to Api trait impl

File: crates/primitives/src/transaction/signature.rs
Comment (fkrause98) [reply crates/primitives/src/transaction/signature.rs:103]: done!

File: crates/rpc/rpc/src/eth/signer.rs
Comment (fkrause98) [reply crates/rpc/rpc/src/eth/signer.rs]: Not sure if this is what you had in mind, but I've added a field to DevSigner that contains a `Secp256k1` :slightly_smiling_face:

File: crates/rpc/rpc/src/eth/signer.rs
Comment (fkrause98) [reply crates/rpc/rpc/src/eth/signer.rs]: Sure!
I've added a new [SignError](https://github.com/paradigmxyz/reth/pull/1665/files#diff-ae449946e8a8fa16f55c2511577cdc5a9efb699b597de667d5b39ea7542610a9R336) enum and a [new variant](https://github.com/paradigmxyz/reth/pull/1665/files#diff-ae449946e8a8fa16f55c2511577cdc5a9efb699b597de667d5b39ea7542610a9R79) for EthApiError

File: crates/primitives/src/transaction/signature.rs
Comment (gakonst) [inline crates/primitives/src/transaction/signature.rs:109]: Do we need to require passing a chain id here by any chance, to respect EIP-155? It is a bit awkward because I think EIP-155 means the signature may end up being larger than 65 bytes if the `v` value exceeds 255 for a large chain id? We've had this conversation before in ethers cc @mattsse

File: crates/rpc/rpc/src/eth/error.rs
Comment (gakonst) [inline crates/rpc/rpc/src/eth/error.rs:342]: are these the standard RPC Error messages that e.g. Geth returns?

File: crates/rpc/rpc/src/eth/signer.rs
Comment (gakonst) [inline crates/rpc/rpc/src/eth/signer.rs:37]: great to see ethers-rs doing the heavy lifting here

File: crates/rpc/rpc/src/eth/signer.rs
Comment (gakonst) [inline crates/rpc/rpc/src/eth/signer.rs:47]: Does the signer need to own the chain id to apply EIP-155? Maybe it's something we do in a follow-up?

File: crates/rpc/rpc/src/eth/signer.rs
Comment (gakonst) [inline crates/rpc/rpc/src/eth/signer.rs:108]: should we replace these with tests from the EIP or from other clients?

**Review by gakonst — COMMENTED** [review]
mostly questions, this looks roughly good to me, pending @mattsse approval

File: crates/rpc/rpc/src/eth/signer.rs
Comment (gakonst) [inline crates/rpc/rpc/src/eth/signer.rs:44]: I think Matt meant using the global `SECP256K1` like we do here https://github.com/paradigmxyz/reth/blob/a0f195590c5b0bc5d7b32a9f449db80aac7018e7/crates/primitives/src/transaction/util.rs#L7?

File: crates/primitives/src/transaction/signature.rs
Comment (mattsse) [reply crates/primitives/src/transaction/signature.rs:109]: in ethers we have the issue that we store the `v: u64` and assumes:

```
        // The u64 to u8 cast is safe because `sig.v` can only ever be 27 or 28
        // here. Regarding EIP-155, the modification to `v` happens during tx
        // creation only _after_ the transaction is signed using
        // `ethers_signers::to_eip155_v`.
        sig[64] = src.v as u8;
```

which is a bit unsafe.

https://github.com/gakonst/ethers-rs/blob/2cead3e11160a48b403d1334eab329d38f10b4fa/ethers-core/src/types/signature.rs#L235-L239

but if properly used the v values is set to  (0|1) + 27:

```
        let v = u8::from(recoverable_sig.recovery_id()) as u64 + 27;
```


https://github.com/gakonst/ethers-rs/blob/2cead3e11160a48b403d1334eab329d38f10b4fa/ethers-signers/src/wallet/mod.rs#L156-L156

we already have (0|1) for `odd_y_parity` so we need to add the 27 when setting the v byte

File: crates/rpc/rpc/src/eth/error.rs
Comment (mattsse) [reply crates/rpc/rpc/src/eth/error.rs:342]: The Unknown account yes, don't know about the other variants.

please add some docs here and check geth compact @fkrause98

File: crates/rpc/rpc/src/eth/signer.rs
Comment (mattsse) [reply crates/rpc/rpc/src/eth/signer.rs:47]: perhaps a reasonable assumption here is that the Signer should do this.

File: crates/rpc/rpc/src/eth/signer.rs
Comment (mattsse) [inline crates/rpc/rpc/src/eth/signer.rs:61]: looks like we already have this function:

https://github.com/paradigmxyz/reth/blob/main/crates/interfaces/src/test_utils/generators.rs#L77-L78

we should move this to primitives imo, so we can reuse this

**Review by mattsse — CHANGES_REQUESTED** [review]

File: crates/rpc/rpc/src/eth/signer.rs
Comment (MegaRedHand) [reply crates/rpc/rpc/src/eth/signer.rs]: Done 👍

File: crates/rpc/rpc/src/eth/error.rs
Comment (gakonst) [reply crates/rpc/rpc/src/eth/error.rs:342]: I am p sure Matt means "compat", i.e. check Geth's errors and ensure there's 1:1 compatibility with the messages

Comment (MegaRedHand) [reply crates/rpc/rpc/src/eth/error.rs:342]: From what I've seen, the only one we have in common is `NoAccount`. Both `CouldNotSign` and `TypedData` are equivalent to a lot of detailed errors in geth (with their messages specifying the exact cause for the error) and would need to be translated from the ones we have in the underlying implementation.

I'll add some docs. Should I make `CouldNotSign` and `TypedData` wrap the more specific errors `secp256k1::Error` and `Eip712Error`?

File: crates/rpc/rpc/src/eth/signer.rs
Comment (MegaRedHand) [reply crates/rpc/rpc/src/eth/signer.rs:46]: This would be needed only for `eth_signTransaction` right?

Comment (mattsse) [reply crates/rpc/rpc/src/eth/signer.rs:46]: yes.

File: crates/primitives/src/transaction/signature.rs
Comment (mattsse) [inline crates/primitives/src/transaction/signature.rs:105]: nit: this should return an 65 array instead.

**Review by mattsse — CHANGES_REQUESTED** [review]
lgtm,

pending nit and conflicts.

great work!

File: crates/rpc/rpc/src/eth/signer.rs
Comment (MegaRedHand) [reply crates/rpc/rpc/src/eth/signer.rs:101]: The data from this test was extracted from _geth_'s `signer/core/signed_data_test.go`. Should I add another one from EIP-712?

**Review by gakonst — APPROVED** [review]
awesome - send it

**Comment by gakonst** [comment]
tysm for this @fkrause98 & @mattsse for the detailed review

**Pull request merged**
